### ISSUE TITLE
New plugin: Mass action. Ideas? Suggestions?

### ISSUE THREAD

**Issue by olleharstedt** [SamMousa/ls3-question-objects-spec#12]
So I'd like to make a plugin to handle mass actions (e.g., edit all question titles in one page) using handsontable. The interface will be excel-like, with save on unfocus for each cell, like in google docs. If anyone has any ideas, suggestions or wishes, please write them down here and I'll try to take it into account.

I will develop for 2.5, and then check how it could be ported to 2.06lts.

**Comment by LouisGac** [comment]
yep: I suggest to do it as a Yii module. So it would be easy to maintain.

**Comment by olleharstedt** [comment]
We don't have any conventions for adding Yii modules as external code. Could you give more details?

**Comment by olleharstedt** [comment]
My suggestions would be to make it a plugin during beta. Once it's stable, it can be included in core.

**Comment by LouisGac** [comment]
My suggestion would be to make it a Yii module. You should give a try, not that hard, and would be modular, and clean. A module can be core, or can be optional, just a question of line of configuration. you don't need to touch its code.

**Comment by olleharstedt** [comment]
@LouisGac That's another discussion: "Stop using plugin system in favor of Yii modules".

**Comment by maziminke** [comment]
Just some thoughts about what may be needed from a user point of view (I have created lots of surveys in the past):
1. Search and replace feature
2. We should also include relevance equation field since that will make updating conditions pretty easy.
3. How about adding a "A backup never hurts" top left to encourage users to export the survey as LSS file before editing?
4. We should also have a kind of "Changes saved" message which shows up for a very short time on each save so that users know that data was already saved.
5. Will "CTRL + Z" work for more than 1 change to undo several edits?
6. A link to the survey logic file (opening in another tab) is also helpful to check adjustments right away.

**Comment by LouisGac** [comment]
well the discussion is that easy :

We're using Yii framework, a beautiful framework, used by dozen of thousands of coders over the world, written by **extremely good coders**, who spend their whole working time about framework questions since years. We're focusing on a survey system, they're focusing on questions such as : "how to handle an extension system".

So, they written an extension system, a very good one, handling any kind of problems an extension system can face, tested by a large community of application makers. It provides a way to build clean extensions, respecting standards such as MVC, thighly bound with Yii. 

Instead of using this system, we're trying to use an "home made plugin system", full of problems, not respecting the most basic standards, not using the framework. **So instead of focusing on our work problems like "_how should we do mass actions_", we quickly come to work on questions such as "how could we improve/correct/fix our home made plugin system".**

That's a mistake. If we rather use the Yii Module system, any Yii coder on the world could write LimeSurvey modules. They could refer to the Yii documentation, instead of trying to understand the logic of our "home made solution".

**Comment by olleharstedt** [comment]
@LouisGac I hear you, but still, this is another topic.

**Comment by LouisGac** [comment]
I created a new topic in this very specific question, and I still answer to your topic : 
you should write a Yii Module to do that.

**Comment by olleharstedt** [comment]
@maziminke A lot of what you suggest depends on the handsontable library. Could you maybe sort your requirements in categories "must have", "should have" and "could have"?

**Comment by SamMousa** [comment]
@olleharstedt I think your problems are solved, just "create a yii module" and all will be fine.

Now in all seriousness:
1. I think a Yii Module would work.
2. Currently LS3 supports loading this but LS2.5 does not, this is not an issue it just means you will need your base plugin class to load the yii module code.
3. We might need to add some functions to the `LimesurveyApi` object that allow you to manipulate the objects you wish to manipulate in a standardized way.

One issue I have is with "exposing" our `ActiveRecord` models. If for example a plugin developer directly uses the `Survey` model (which we can't technically prevent but we should offer an alternative) then he / she might wrongfully assume that all it's methods are part of the plugin API. In which case if we change functions we use internally this plugin might break.

My suggestion would be to change the models in the core a bit to allow for a separate plugin AR model backed by the same table, this might need some more thinking but consider this:
1. A `ls\models\BaseSurvey` class for the survey table.
2. A `ls\api\models\Survey` class extending (1) that we can keep stable for plugin developers.
3. A `ls\models\Survey` class extending (1) that we use internally and where we can add functions we need internally.

There are plenty of other solutions I think, but the most important thing is that we clearly mark what is and what is not part of the "public API". Without this any minor upgrade might break plugins which is not a good environment for plugin developers or end users.

**Comment by olleharstedt** [comment]
Is the pain worth the gain? How would users like Mazi test a module?

**Comment by LouisGac** [comment]
All those questions are "**framework**" questions, not "**survey application**" questions.
We should use the framework, instead of pretending making a better system than the Yii dev team.

**Comment by olleharstedt** [comment]
OK, let's keep all those questions in the other thread. Here we'll just assume it's a plugin for now. Thank you.

**Comment by SamMousa** [comment]
@olleharstedt I don't think the pain is worth the gain short term.
If you code it nicely you could always refactor it for when Yii modules are supported.
Note that long before yii existed PHP developers could make modular code and that is what this is about.

In this case since your code actually will do full page rendering (view layer), handle user requests (controller layer) and store data via the API (model layer) a module would provide you with an easy structure though.

That said, altering LS2.5 to support that might be too big of a change.

So if you create clean code then it will be fine regardless of being a Yii module or not. When LS3 supports real modules you can just remove some code from your plugin (like request routing) that will then be done by the core instead of custom built for your plugin

**Comment by LouisGac** [comment]
LS2.5 is based on Yii, so it supports module.
True question is : why the home made plugin system as been written instead of using the framework?

**Comment by olleharstedt** [comment]
@LouisGac Please keep that discussion in the designated thread.

**Comment by LouisGac** [comment]
I do. My point of view is that the Mass actions are an important feature so they should not use the "home made plugin system".

**Comment by SamMousa** [comment]
We are discussing the API @olleharstedt would need for this and the user requirements such a feature might have.
Just shouting "use Yii modules" is not an answer to either of those questions.

**Comment by LouisGac** [comment]
"if anyone has any ideas, suggestions or wishes, please write them down here and I'll try to take it into account."
That the thread. 
My wish: don't use the home made plugin system, which is not maintained, which as unresolved bugs (like for the LDAP plugin), to do it.

**Comment by SamMousa** [comment]
@LouisGac Yii 1 isn't supported anymore either. Regardless of that the plugin system itself has no real bugs that I'm aware of. It might have limitations.
Note that a broken plugin can still happen if it's a Yii module.

**Comment by maziminke** [comment]
@olleharstedt "_Could you maybe sort your requirements in categories "must have", "should have" and "could have"?_"
-> I have added that directly at my list.

**Comment by LouisGac** [comment]
Yii1 is supported :  
 https://github.com/yiisoft/yii/releases

The LimeSurvey "home made plugin system" is not supported. It has been used for very important features like "login", and some of the plugins used for kernel are still not working. 

If the LDAP plugin is not fixed by the plugin code owner, I'll have to redo it myself. Then I'll probably redo it using Yii Module system.

**Comment by SamMousa** [comment]
@LouisGac you don't make those decisions, @c-schmitz  does.

**Comment by olleharstedt** [comment]
@maziminke Great.

**Comment by trivender** [comment]
It is a revolutionary idea and love to have this feature. Save on un-focus is cool but more important is would be able to copy multiple rows from excel/open-office and paste. It takes lots of time when you have change relevance condition of all question in a group.  A similar plug will also be helpful for translating survey in different languages . say "Translate this group" ... Once clicked shows input fields for all question texts and answer labels. We can then quickly copy/paste lists from other application and save in one go.

**Comment by olleharstedt** [comment]
Copy/paste depends on the JS library. Seems to work, try it out here: https://handsontable.com/examples.html?headers

**Comment by Shnoulle** [comment]
The final plugin/modules is not the question i think.

For me : we need a way to allow external dev to
- Add a menu in all/some menu
- Add a admin page with content according too 'sid','qid',gid',.... 

After : we need to add a system for autoupdating the plugin , auto download install etc...

Actually : anyone can include a new controller in application/controller : it's not difficult. But it's not the good way (i think).

**Comment by olleharstedt** [comment]
Maybe we can use code from Louis's ComfortUpdate code for plugin autoupdating.

Adding new third-party controllers should be in a specific directory. Personally, I don't see an immediate need to add another system for third-party code. There are more important issues.

**Comment by Shnoulle** [comment]
AutoUpdate/AutoLoad : It's one of the power of wordpress and a lot of CMS. Piwik have it too.

PS : i think we can have different repo url

**Comment by Shnoulle** [comment]
For 6 : Could have: A link to the survey logic file (opening in another tab) is also helpful to check adjustments right away. : NO : better solution is a link "Validate THIS question" at each line : open a dialog . SOmething like "Validate email model" or doing it in ajax and not on iframe (but iframe is quickest).

**Comment by olleharstedt** [comment]
@Shnoulle Denis, how would I do to add a script file from a plugin that is loaded on every admin page? Do we have an event for that? Like `beforeAdminPageLoad`, then plugin can register script file.

**Comment by Shnoulle** [comment]
I add beforeControllerAction for this. Before : you can use same event than https://framagit.org/SondagePro-LimeSurvey-plugin/hideUpdateAdmin

**Comment by olleharstedt** [comment]
Thanks, will try.

**Comment by olleharstedt** [comment]
Recent changes: Add question group; confirm message after save is complete.
